plugins/inputs/jboss/jboss.go

```diff
@@ -0,0 +1,1312 @@
+	ContextRoot       string                 `json:"context-root"`
+	ExpiredSessions   string                 `json:"expired-sessions"`
+	MaxActiveSessions string                 `json:"max-active-sessions"`
+	SessionsCreated   string                 `json:"sessions-created"`
+	Servlet           map[string]interface{} `json:"servlet"`
+}
+
+// JBoss the main collectod struct
+type JBoss struct {
+	Servers []string
+	Metrics []string
+
+	Username string
+	Password string
+
+	Authorization string
+
+	ResponseTimeout internal.Duration
+
+	tls.ClientConfig
+
+	client HTTPClient
+}
+
+// HTTPClient HTTP client struct
+type HTTPClient interface {
+	// Returns the result of an http request
+	//
+	// Parameters:
+	// req: HTTP request object
+	//
+	// Returns:
+	// http.Response:  HTTP respons object
+	// error        :  Any error that may have occurred
+	MakeRequest(req *http.Request) (*http.Response, error)
+
+	SetHTTPClient(client *http.Client)
+	HTTPClient() *http.Client
+}
+
+// RealHTTPClient the HTTP client handler
+type RealHTTPClient struct {
+	client *http.Client
+}
+
+// MakeRequest do an HTTP request
+func (c *RealHTTPClient) MakeRequest(req *http.Request) (*http.Response, error) {
+	return c.client.Do(req)
+}
+
+// SetHTTPClient set http client
+func (c *RealHTTPClient) SetHTTPClient(client *http.Client) {
+	c.client = client
+}
+
+// HTTPClient return the HTTP client
+func (c *RealHTTPClient) HTTPClient() *http.Client {
+	return c.client
+}
+
+var sampleConfig = `
+  # Config for get statistics from JBoss AS
+  servers = [
+    "http://localhost:9090/management",
+  ]
+  
+  ## Username and password
+  username = ""
+  password = ""
+  
+  ## authorization mode could be "basic" or "digest"
+  authorization = "digest"
+
+  ## Optional SSL Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+  ## Metric selection
+  metrics =[
+	"jvm",
+	"web", 		# Handles both EAP <=6.X/AS <=7.X and EAP >=7.X/Widlfly > 8
+	"deployment",
+	"database",
+	"transaction",
+	"jms",
+  ]
+`
+
+// SampleConfig returns a sample configuration block
+func (*JBoss) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description just returns a short description of the JBoss plugin
+func (*JBoss) Description() string {
+	return "Telegraf plugin for gathering metrics from JBoss AS"
+}
+
+// Gather Gathers data for all servers.
+func (h *JBoss) Gather(acc telegraf.Accumulator) error {
+
+	if h.ResponseTimeout.Duration < time.Second {
+		h.ResponseTimeout.Duration = time.Second * 5
+	}
+
+	if h.client.HTTPClient() == nil {
+		tlsCfg, err := h.ClientConfig.TLSConfig()
+
+		if err != nil {
+			return err
+		}
+		tr := &http.Transport{
+			TLSClientConfig: tlsCfg,
+		}
+		client := &http.Client{
+			Transport: tr,
+			Timeout:   h.ResponseTimeout.Duration,
+		}
+		h.client.SetHTTPClient(client)
+	}
```

Comment:
Please move this to an `Init() error` function that is automatically called by telegraf once all config parameters are parsed and set but before the first gather. This avoids doing the same init code over and over and also helps the reader to understand that it is intended to be done once.